sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,396 @@
+                                                        TimeSpan? removeAfterDuration = default)
+        {
+            if (string.IsNullOrEmpty(partitionId))
+            {
+                return new PooledProducer(this);
+            }
+
+            var identifier = Guid.NewGuid().ToString();
+            var item = Pool.GetOrAdd(partitionId, id => new PoolItem(partitionId));
+
+            // A race condition at this point may end with CloseAsync called on
+            // the returned PoolItem if it had expired. The probability is very low and
+            // possible exceptions should be handled by the invoking methods.
+
+            if (item.PartitionProducer?.IsClosed == true || !item.ActiveInstances.TryAdd(identifier, 0))
+            {
+                identifier = Guid.NewGuid().ToString();
+                item = Pool.GetOrAdd(partitionId, id => new PoolItem(partitionId));
+                item.ActiveInstances.TryAdd(identifier, 0);
+            }
+
+            item.ExtendRemoveAfter(removeAfterDuration);
+
+            return new PooledProducer(this, partitionId, cleanUp: () =>
+            {
+                if (Pool.TryGetValue(partitionId, out PoolItem pooledItem))
+                {
+                    pooledItem.ExtendRemoveAfter(removeAfterDuration);
+                }
+
+                // If TryRemove returned false the PoolItem would not be closed deterministically
+                // and the ExpirationTimer callback would eventually remove it from the
+                // Pool leaving to the Garbage Collector the responsability of closing
+                // the TransportProducer and the AMQP link.
+
+                item.ActiveInstances.TryRemove(identifier, out _);
+
+                // The second TryGet runs after the extension would have been seen, so it
+                // is intended to be sure that the item wasn't removed in the meantime.
+
+                if (!Pool.TryGetValue(partitionId, out _) && !item.ActiveInstances.Any())
+                {
+                    return item.PartitionProducer?.CloseAsync(CancellationToken.None);
+                }
+
+                return Task.CompletedTask;
+            });
+        }
+
+        /// <summary>
+        ///   Creates a <see cref="TransportProducer" /> matching the partition id passed as input.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The unique identifier of a partition associated with the Event Hub.</param>
+        /// <param name="connection">The <see cref="EventHubConnection" /> connection to use for communication with the Event Hubs service.</param>
+        /// <param name="retryPolicy">The policy to use for determining retry behavior for when an operation fails.</param>
+        ///
+        /// <returns>A <see cref="TransportProducer" /> matching the partition id passed as input.</returns>
+        ///
+        public virtual TransportProducer GetTransportProducer(string partitionId = default,
```

Comment:
Based on the summary, should this be called `CreateTransportProducer`?    If we take the connection and retry policy as part of the pool's constructor, does that let us make this private and have the pool own responsibility for creating a transport producer when a new pooled producer is needed?

Reply:
Good point. I momentarily changed the summary and the implementation based on your great suggestion above. I am super happy to revisit is as you find most appropriate.